### PR TITLE
Add profiles proto

### PR DESCRIPTION
**Pull request by petethepig**
This PR adds a new profiles signal. This is currently very much in a work-in-progress / seeking feedback kind of stage.

Relevant documents / previous discussions:
* [OTEL Profiling Vision](https://github.com/open-telemetry/oteps/blob/main/text/profiles/0212-profiling-vision.md)
* [Profiling SIG Meeting Notes](https://docs.google.com/document/d/1D-1MbT3MWpKM61JgVFx_TQi4vqWLLIZp6biK5MoHafI/edit?usp=sharing)
* [Draft PR in OTEL-collector](https://github.com/petethepig/opentelemetry-collector/pull/1)

---

The overall design is based on/heavily inspired by:
* other OTEL signals
* [Felix Geisendörfer's draft for a profiles proto](https://github.com/felixge/opentelemetry-proto/pull/1)
* [Anton Kolesnikov's draft for a profiles proto](https://gist.github.com/kolesnikovae/d61f6b668fc093eb40ee7e7a2a5afb64)
* [pprof proto](https://github.com/google/pprof/blob/main/proto/profile.proto)

This PR consists of 2 parts:
* The `profiles.proto` and `profiles_service.proto` which are mostly boilerplate and more or less finished from my perspective
* 3 alternative `Profile` implementations, still very much work-in-progress
  * `pprof` — included as a reference and as a baseline for future benchmarks
  * `denormalized` — most messages are embedded, no string table, a lot of potential for duplication => easier to work with, potentially larger payloads
  * `normalized` — most messages are linked using a foreign key, includes a string table, not much duplication => harder to work with, potentially smaller payloads

### Things that are currently supported:
* support for OTEL-compatible attributes (labels) on Profile and Samples level. Meaning you can label both whole profiles and individual stack traces
* linking with traces. TODO: do we need linking with other signals?
* support for multiple profile types in one `Profile` (similar to how it's implemented in pprof and used in go memory profiles via multiple `value` values)
* ability to specify `AggregationTemporality` for each `SampleType` (e.g something like `alloc_space` in go is cumulative)
* optional per-sample timestamps
* support for embedded original profile (`original_payload` field)

### Design TODOs / Areas where feedback would be most valuable:
* The biggest thing is we need to decide which one of these approaches (`denormalized` / `normalized` / mix of the two) we're going to take. I personally think that the optimal end product should be closer to `normalized` implementation, especially considering that on-the-wire efficiency is a [particularly important part of OTEL's profiling vision](https://github.com/open-telemetry/oteps/blob/main/text/profiles/0212-profiling-vision.md#standardize-profiling-data-model-for-industry-wide-sharing-and-reuse). But we still need to do some benchmarks to confirm that that would actually be significantly more efficient. 
* Figure out what to do with pprof messages like `Mapping` — I'm not sure if we need those
* Improve `SampleType` message. Currently `Type` and `Unit` are strings. We might want to turn them into enums, or standardize them some other way, e.g making those a part of [OTEL semantic conventions](https://opentelemetry.io/docs/concepts/semantic-conventions/)
* Things I'm forgetting / not thinking about

### Implementation TODOs / next steps
* get a corpus of some real production profiling data
* run benchmarks with this corpus of data -> fine-tune various aspects of the format
* implement a few simple OTEL processors -> fine-tune various aspects of the format
* create a reference implementation for a client -> fine-tune various aspects of the format
* create a reference implementation for a server -> fine-tune various aspects of the format
* verify that the spec works with those reference implementations -> fine-tune various aspects of the format

## Reference Diagrams

For reference, here are graph representations of other signals:

<details>

  <summary>metrics</summary>
  
metrics dot
</details>

<details>
  <summary>logs</summary>
  
logs png dot

</details>

<details>
  <summary>traces</summary>

  traces dot
</details>

---

And these are graph representations of alternatives formats under consideration:

<details>
  <summary>pprof for reference</summary>
  
pprof dot

</details>

<details>
  <summary>denormalized</summary>

denormalized dot
</details>

<details>
  <summary>normalized</summary>
  

normalized dot
</details>

---

/cc @tigrannajaryan @felixge @jhalliday @jsuereth @christos68k

### CODE REVIEW

**Comment by tigrannajaryan** [comment]
@petethepig we have just discussed this in the TC meeting.

We would like to give you a separate repo in Otel github org, a fork of this one, where you can experiment freely with the profile protos. We will give you approving/merging rights on this new repo, so that you can also iterate faster and don't need to wait for approvals/merging from OTLP maintainers.

You can also fork Collector and easily change it to consume proto file files from your fork instead of this repo and do your experimentation in the Collector too.

When the experiments are concluded you can do a cleanup on your fork and then we will merge back the finalized version of profile protos back into this repo.

What do you think?

cc @jsuereth

File: opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto
Comment (cyriltovena) [inline opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto:26]: I wonder if those should be int32/uint32 I don't think we'll overflow this specially since they refer slice/array indexes.

This is not super important for the wire format since proto will varint encode it but for the collector this will allow it to buffer more as the memory footprint will be lower per profiles.

File: opentelemetry/proto/profiles/v1/profiles.proto
Comment (cyriltovena) [inline opentelemetry/proto/profiles/v1/profiles.proto:102]: nit: wouldn't be better to have a duration instead. I also wonder what happens for cumulative profile.

File: opentelemetry/proto/profiles/v1/profiles.proto
Comment (cyriltovena) [inline opentelemetry/proto/profiles/v1/profiles.proto:130]: I'm waiting to see some benchmark, but I do think the `normalized` is the way to go.

With a lot of samples (we can expect ~10k per profiles) it avoid duplications. It will most likely have a smaller memory footprint and faster marshalling/unmarshalling.

It's also closer to pprof which is quite popular for profiling data.

File: opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto
Comment (florianl) [inline opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto:146]: For postprocessing it might be great for the backend to have additional (optional) information. Additional information can be something like, if this mapping belongs to some interpreter.

Maybe something like:
```
repeated Label label = 11;
```

**Review by florianl — COMMENTED** [review]
Thanks for the work @petethepig && @felixge  :pray:

File: opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto
Comment (thomasdullien) [inline opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto:27]: Is there a good reason to use gzip, when pareto-improved compression methods are available? (e.g. algorithms that outperform gzip compression at less CPU usage at every level?)

File: opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto
Comment (cyriltovena) [reply opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto:110]: I thought this ID would be to reference locations inside this profile. Are you thinking about global addressing ?

File: opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto
Comment (thomasdullien) [inline opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto:104]: Display tools may not have the ability to find the beginning of the instruction unless the "in the middle" above is specified precisely.

File: opentelemetry/proto/profiles/v1/alternatives/denormalized/denormalized.proto
Comment (thomasdullien) [inline opentelemetry/proto/profiles/v1/alternatives/denormalized/denormalized.proto:95]: We need to specify *exactly* how deep into the call instruction the address is allowed to be, display tools do not have access to the assembly bytes and hence cannot do more than simple arithmetic to find the beginning of the instruction.

File: opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto
Comment (thomasdullien) [inline opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto:125]: Is this an index into the other array?

Comment (thomasdullien) [reply opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto:125]: Perhaps we should rename `_id` to `_index`?

File: opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto
Comment (thomasdullien) [inline opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto:85]: Could the comment be clarified?

File: opentelemetry/proto/profiles/v1/profiles.proto
Comment (jhalliday) [reply opentelemetry/proto/profiles/v1/profiles.proto:130]: @cyriltovena Explain the math behind that 10k, please?

File: opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto
Comment (thomasdullien) [inline opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto:204]: Is that an ID or an index?

File: opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto
Comment (thomasdullien) [inline opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto:211]: Is that an ID or an index?

File: opentelemetry/proto/profiles/v1/profiles.proto
Comment (thomasdullien) [inline opentelemetry/proto/profiles/v1/profiles.proto:2]: Cut & paste issue about Copyright year.

**Review by thomasdullien — CHANGES_REQUESTED** [review]
My instinct is also for the normalized.

I'd like to make sure that we always use the suffix `_index` for things that are array indices, and reserve `_id` for things that are likely hashes or other pseudorandom values. I'd also *strongly* push for using 128-bit values for `_id`s, always and everywhere, lest we create issues for everybody when collisions arise. 64-bit values collide randomly at data quantities that are absolutely the norm these days.

File: opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto
Comment (florianl) [reply opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto:110]: Ah yes - this comment would also fit to `Location.id` just like https://github.com/open-telemetry/opentelemetry-proto/pull/488#discussion_r1246639511.

One of the motivations is really to avoid collisions in global addressing. Even if the same code is compiled, differences in the build systems/compilers are important and can make a difference.

File: opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto
Comment (cyriltovena) [reply opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto:204]: index in that profile.

File: opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto
Comment (petethepig) [inline opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto:27]: embed these in attributes?

File: opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto
Comment (petethepig) [inline opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto:26]: Experiment with storing these as a tree

File: opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto
Comment (petethepig) [inline opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto:35]: byte-seconds might overflow when aggregated

Comment (petethepig) [reply opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto:35]: should not be fixed64

File: opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto
Comment (petethepig) [inline opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto:70]: Florian: should stay for profiles that don't have symbols

**Comment by petethepig** [comment]
TODO: add a design decision log?

**Comment by jsuereth** [comment]
@tigrannajaryan I created https://github.com/open-telemetry/opentelemetry-proto-profile for the profiling group to own and experiment with.

So far I only add @Rperry2174 as a maintainer.  I would add @petethepig but he's not a member of the OpenTelemetry group yet (please follow this: https://github.com/open-telemetry/community/blob/main/community-membership.md#community-membership, I can vouch for you as one of your two)

File: opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto
Comment (petethepig) [reply opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto:211]: In pprof it is an ID. And there is some confusion right now because usually IDs and indexes are identical, but not always. The plan for the new spec is that we're going to remove ids and rely on indexes. This way there's one way to address objects and so there's going to be:
* less confusion
* better performance (array lookups are faster than search for ID)

File: opentelemetry/proto/profiles/v1/profiles.proto
Comment (petethepig) [reply opentelemetry/proto/profiles/v1/profiles.proto:130]: @christos68k Yeah, pprof was included as a reference so that we'd be able to easily compare it to other ones. Long term plan is to keep just one of these 3 formats, and based on the discussions so far it seems like it will be based on the `normalized` version.

File: opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto
Comment (petethepig) [reply opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto:86]: @aalexand Can you provide the list of combinations that are valid? I remember you mentioned that not all 16 combinations are valid, but also 4-way enum won't cut it either, so maybe we can make this an enum with more than 4 but less than 16 combinations.

Comment (aalexand) [reply opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto:86]: I think the combinations encountered in practice are these - see below.

| has_functions  | has_filenames | has_line_numbers | has_inline_frames | Description |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| true | false | false | false | A C++ / native binary with just symbol table (.symtab), no DWARF. |
| true | true | true | true | A C++ / native binary symbolized using profiler-friendly (i.e. at least `-gmlt`) DWARF. |
| true | true | true | false | Full symbol information but no distinction of inline frames - e.g. from Java. Can also correspond to an exotic case of using separate file DWARF, but not providing the *.dwp file. |
| false | false | false | false | Unsymbolized mapping |

I would appreciate if someone also familiar with the subject could verify this from their perspective.

**Comment by tigrannajaryan** [comment]
@petethepig please re-create this PR in https://github.com/open-telemetry/opentelemetry-proto-profile

I am going to close this PR and you can continue discussing it in the new repo.

File: opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto
Comment (petethepig) [reply opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto:86]: @aalexand thank you! I was trying to make an enum out of this, and here's what I have so far:
```proto
  enum SymbolicInfo {
    SYMBOLIC_INFO_UNSPECIFIED = 0;
    SYMBOLIC_INFO_FULL = 1;
    SYMBOLIC_INFO_FUNCTIONS_ONLY = 2;
    SYMBOLIC_INFO_NO_INLINE_FRAMES = 3;
  }
```

As I was doing this one thing I thought of is do we need these booleans / enums at all? My understanding is that these booleans are useful when combining a profile without symbols with symbols information (e.g in DWARF format). But I would assume that when you have symbols in some different format (let's say DWARF), you already know if it has functions / line numbers / etc. So it feels like maybe we don't need these booleans / enums. Unfortunately, my understanding of all of this is still pretty limited, so I would love your input on this.

---

Btw, I checked google/pprof repo, and it looks like out of these 4 booleans, the only one that's meaningfully used is [HasFunctions one](https://github.com/search?q=repo%3Agoogle%2Fpprof+HasFunctions&type=code), and even then there's only one mention (`symbolz.go`). The other fields are just copied back and forth but not actually used it seems like.

cc @korniltsev

File: opentelemetry/proto/profiles/v1/profiles.proto
Comment (petethepig) [reply opentelemetry/proto/profiles/v1/profiles.proto:102]: @cyriltovena 

I have 2 questions:
* Can you elaborate on why it would be better to have a duration? 
* Why is this going to be different for cumulative profiles? I'm not following

Btw I copied this from trace proto, so there's value in keeping like this for consistency.

File: opentelemetry/proto/profiles/v1/profiles.proto
Comment (petethepig) [reply opentelemetry/proto/profiles/v1/profiles.proto:75]: @florianl 

I copied this part from other signals, like traces or logs — so if this is truly a limitation then I think it should be discussed in the broader OTEL context.

I don't think it is a limitation though — [instrumentation scope describes 2 things](https://github.com/open-telemetry/opentelemetry-proto/blob/3decfa005db2885341c03843470681646d5ab239/opentelemetry/proto/common/v1/common.proto#L69-L81):
* instrumentation scope name and version
* attributes

I think it is intended that scope name and version would be the same for system-wide profile. RE attributes, the format in its current state is very flexible when it comes to assigning attributes, for example, clients can create multiple separate profiles within one `ScopeProfiles`, each with it's own set of attributes, allowing for "per application" attributes within one instrumentation scope.

File: opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto
Comment (petethepig) [reply opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto:146]: https://github.com/petethepig/opentelemetry-proto/commit/980fc893bf5c0ebfe8bc9e77af2b36971f6ba581

I added it to `Location` as well. My logic here is that the same argument can be made for adding it to `Location`, so I figured it makes more sense to include it there. I made it a reference to an `AttributeSet`, same way the linking is done from `Sample`.

@florianl what do you think?

File: opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto
Comment (petethepig) [reply opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto:85]: I think this is sample rate basically. I'm not going to edit this one because this is copied from pprof spec. I will make sure we document it well in the final otel proto spec though.

File: opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto
Comment (petethepig) [reply opentelemetry/proto/profiles/v1/alternatives/pprof/pprof.proto:27]: This is pprof proto and I only included pprof proto as an example.

Either way, from my perspective, compression is something that is broadly applicable to the whole OTEL / OTEL collector universe and therefore we should bring these kinds of things up to the broader OTEL collector maintainers group.

File: opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto
Comment (petethepig) [reply opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto:103]: This comment comes from pprof proto. My understanding is that this comment refers to the limitation of some profilers where they are not always able to get exact instruction address for a given location. So it's less about what the implementer should put into this field and more of a word of caution for implementers of display tools. I might be totally wrong though — this is getting a little outside of my area of expertise, so I will need help from someone.

File: opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto
Comment (petethepig) [reply opentelemetry/proto/profiles/v1/alternatives/normalized/normalized.proto:125]: Yeah, this makes a lot more sense — changed it to `_index` and `_indices` everywhere.

**Pull request closed without merging**
